feat(video): create Video component

kaseyvee opened the pull request:
## Changes
- create video component
- refactor controls to work with video

## Screenshots/Videos
### Designs
#### Desktop
<img width="673" alt="Screenshot 2024-04-11 at 10 43 58 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/eb4c62e6-54a3-4ca6-81bb-7ff58da04311">
<img width="673" alt="Screenshot 2024-04-11 at 10 44 08 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/d3d68109-4a55-46f3-bc38-520c3f41c8f5">

#### Mobile
<img width="360" alt="Screenshot 2024-04-11 at 10 44 23 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/c56b2bb3-e19c-403a-920f-2efeb1e9bcf8">
<img width="360" alt="Screenshot 2024-04-11 at 10 44 48 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/c3356ca3-21e3-4853-b38e-4d3fc0aa095c">
<img width="360" alt="Screenshot 2024-04-11 at 10 44 54 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/07af95d0-5b14-44f1-896a-6ca1e9dbeb26">
<img width="360" alt="Screenshot 2024-04-11 at 10 45 02 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/3e54a28b-6d0a-4746-9ea4-776b1e134eb9">

### Storybook
<img width="1028" alt="Screenshot 2024-04-11 at 10 55 39 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/a1d9c636-1ab2-4fad-abda-f67e139b4670">
<img width="1028" alt="Screenshot 2024-04-11 at 10 55 19 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/2aada2b2-3590-4a4e-9ef6-83111f1b7a20">

#### Light mode
Default
<img width="807" alt="Screenshot 2024-04-15 at 11 14 37 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/8d9bbe71-171d-428e-8b62-6104d98b6bd4">

With captions
<img width="807" alt="Screenshot 2024-04-15 at 11 15 15 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/cb0b99a2-3002-4cb0-86e0-962d66f3d99c">

With poster
<img width="807" alt="Screenshot 2024-04-15 at 11 15 29 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/6446ed6c-1570-496a-8219-c069f15d729e">

Error
<img width="1028" alt="Screenshot 2024-04-11 at 10 55 54 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/3f6e94a3-8eda-4bc3-ae77-2514cae924e6">

Mobile
<img width="540" alt="light-mobile" src="https://github.com/rustic-ai/ui-components/assets/111031789/de89290a-c90d-4094-a226-d1c6f74bb063">

#### Dark mode
Default
<img width="814" alt="Screenshot 2024-04-15 at 11 12 44 AM" src="https://github.com/rustic-ai/ui-components/assets/111031789/6a0044c1-0ff8-4f24-874c-4bac3595e02d">

Mobile
<img width="540" alt="dark-mobile" src="https://github.com/rustic-ai/ui-components/assets/111031789/8ae0dfeb-c1f5-4faa-8829-ac9351ec7241">

#### Videos
Desktop Fullscreen
https://github.com/rustic-ai/ui-components/assets/111031789/e0ad61a1-f119-4c3e-adb9-068c3fe5f0f7

Desktop Picture-in-picture
https://github.com/rustic-ai/ui-components/assets/111031789/2f22fe6d-90b7-4c4e-b7e4-a2109d7f5c26

Mobile Fullscreen
https://github.com/rustic-ai/ui-components/assets/111031789/36a9f7fd-4652-4d72-be52-a6e4e32f56b8

Shiti commented on src/components/media/controls/controls.tsx:
```diff
@@ -9,23 +9,41 @@ import IconButton from '@mui/material/IconButton'
+    | 'pictureInPicture'
+    | 'pictureInPictureExit'
+    | 'fullscreen'
+    | 'fullscreenExit'
```
I think it would be good to add the captions ones here too

Shiti commented on src/components/media/controls/controls.tsx:
```diff
@@ -9,23 +9,41 @@ import IconButton from '@mui/material/IconButton'
+    | 'fullscreen'
+    | 'fullscreenExit'
   className?: string
+  color?: string
```
why do we need to pass the color as prop? can we not set it to 'primary' always ?

kaseyvee replied:
The colours change based on the component as well as the viewport, so I was trying to think of a way to control this.

Shiti replied:
No, the colors should not change by viewport. Please keep this to primary

Shiti reviewed (commented):
I'm thinking we should split the controls file into 3 - mediaIcons component, common(controls common to audio and video) and videocontrols. what do you think ?

Shiti reviewed (commented):
Please update the commit message to clearly indicate what is being changed. Also, for tests, I would recommend using a local video file as that would help speed things up

lyjeileen commented:
Could you hide the control bar when the mouse has not moved for a few seconds?

lyjeileen commented on src/components/types.ts:
```diff
@@ -136,11 +136,18 @@ export interface TableFormat extends DataFormat {
   transcript?: string
 }
+
+export interface AudioFormat extends MediaFormat {}
```
Question: Is it necessary to add `AudioFormat` when it's exactly same as `MediaFormat`? Do we want to introduce a new type just for the semantic reason?

kaseyvee replied:
Semantics for clarity but I'm also thinking in the future there could possibly be audio-only fields.

kaseyvee commented:
> I'm thinking we should split the controls file into 3 - mediaIcons component, common(controls common to audio and video) and videocontrols. what do you think ?

@Shiti This would make it easier to maintain. Will still have to think about how the controls in the settings menu will work.

Shiti commented on src/components/media/controls/commonControls.tsx:
```diff
@@ -1,31 +1,26 @@
-  isTranscriptShown: boolean
-  setIsTranscriptShown: () => void
+interface Toggle {
+  active: boolean
```
```suggestion
  isActive: boolean
```

I do think that this interface is too generic and does not explain anything.

kaseyvee replied:
What about:
```
interface ToggleButton {
  isSomeState: boolean
  setIsSomeState: () => void
}
```

Shiti commented on src/components/types.ts:
```diff
@@ -136,11 +136,18 @@ export interface TableFormat extends DataFormat {
 export type TableData = TableFormat & Updates<TableFormat>
 
-export interface AudioFormat extends DataFormat {
+export interface MediaFormat extends DataFormat {
```
Please ensure that the docs for each field refer to `media` instead of `audio`

kaseyvee commented on src/components/media/controls/commonControls.tsx:
```diff
@@ -170,60 +143,75 @@ export function VolumeSettings(props: MediaControls) {
-  const Icon = props.isTranscriptShown
-    ? KeyboardArrowUpIcon
-    : KeyboardArrowDownIcon
+export function CaptionsToggle(props: Toggle) {
```
I figured since we're using it in both the `Sound` and `Video` components for now until the `Video` settings menu design flow was finalized, since it seems like playback rate, captions, etc will later be handled there.

Laurendragonscale commented:
We should consider adding a lot more contrast behind the video controls. Maybe even a solid colour like this:
<img width="842" alt="Screenshot 2024-04-11 at 2 57 37 PM" src="https://github.com/rustic-ai/ui-components/assets/142948108/a9d15df6-c8d5-4f34-b4e1-a1e0c3f1e28a">

lyjeileen commented:
Hi! I found a bug in the mobile version. Here are the steps to reproduce it:
- Click `play` button (It switched to the full screen mode)
- Click the `full screen` toggle (It switched back to the small screen)
- Click `play` button again => Nothing happened
- Click `play` button again (It switched to the full screen mode)

lyjeileen commented:
Should the `video` folder be in the `media` folder?

kaseyvee commented:
> Hi! I found a bug in the mobile version. Here are the steps to reproduce it:
> 
> * Click `play` button (It switched to the full screen mode)
> * Click the `full screen` toggle (It switched back to the small screen)
> * Click `play` button again => Nothing happened
> * Click `play` button again (It switched to the full screen mode)

@lyjeileen Addressed. Initially spoke with @RenataDzotova about making this immediately go to fullscreen on mobile since it's not obvious you have to go to fullscreen for the controls. I've updated to be back to the original design. So clicking play now simply plays the video without going fullscreen and added back the fullscreen button.

kaseyvee commented on src/components/video/video.tsx:
```diff
@@ -0,0 +1,343 @@
+        videoRef.current &&
+        videoContainerRef.current &&
+        isMobile &&
+        !isFullscreen
```
After addressing [this](https://github.com/rustic-ai/ui-components/pull/72#issuecomment-2050721647), this is no longer relevant

Shiti commented on src/components/media/audio/sound.cy.tsx:
```diff
@@ -95,7 +95,7 @@ describe('Sound', () => {
         />
       )
-      cy.get('[data-cy=captions-toggle]').click()
+      cy.get('[data-cy=captionsOn-button]').click()
```
@kaseyvee this was not resolved

Shiti commented:
@kaseyvee when transcript is shown in full screen mode, the other controls are not visible even on hover. We have to hide transcript for any other controls to be visible.

Shiti commented:
When showing picture-in-picture, I think it should show the poster /first frame at the original location.

Shiti commented on src/components/media/controls/videoControls.tsx:
```diff
@@ -0,0 +1,54 @@
+    ? 'pictureInPictureExit'
+    : 'pictureInPicture'
+
+  function handlePictureInPicture() {
+    if (isPictureInPicture) {
+      document.exitPictureInPicture()
+    } else {
+      videoElement.requestPictureInPicture()
+    }
+  }
```
Both `document.exitPictureInPicture()` and `videoElement.requestPictureInPicture()` return Promise. Use that instead of adding event handlers below.

kaseyvee replied:
I've left the `onleavepictureinpicture` handler for when users exit from the pip window.

Shiti commented on src/components/media/controls/videoControls.tsx:
```diff
@@ -0,0 +1,54 @@
+
+  function handleFullscreen() {
+    if (isFullscreen) {
+      document.exitFullscreen()
+    } else {
+      props.element.requestFullscreen()
+    }
```
Please update this to handle the returned Promise.

kaseyvee commented:
> @kaseyvee when transcript is shown in full screen mode, the other controls are not visible even on hover. We have to hide transcript for any other controls to be visible.

@Shiti Not quite sure how the designs are meant to handle showing the transcript on fullscreen with the controls. I could make them appear like this with a max height of 25% viewport height:
<img width="492" alt="Screenshot 2024-04-12 at 1 13 55 PM" src="https://github.com/rustic-ai/ui-components/assets/111031789/00bea4bd-4995-4795-938f-51e62064cfab">

@Laurendragonscale @RenataDzotova @lyjeileen What do you guys think?

Shiti commented:
on desktop, the transcript comes below the controls so I would assume that the same happens on mobile as well.

Shiti commented on src/components/media/controls/videoControls.tsx:
```diff
@@ -0,0 +1,92 @@
+      document
+        .exitPictureInPicture()
+        .then(() => {
+          props.onError('')
```
We shouldnt be calling `onError` in case of successful response
```suggestion
```

Shiti commented on src/components/media/controls/videoControls.tsx:
```diff
@@ -0,0 +1,92 @@
+      videoElement
+        .requestPictureInPicture()
+        .then(() => {
+          props.onError('')
```
We shouldnt be calling `onError` in case of successful response
```suggestion
```

Shiti commented on src/components/media/controls/videoControls.tsx:
```diff
@@ -0,0 +1,92 @@
+      document
+        .exitFullscreen()
+        .then(() => {
+          props.onError('')
```
We shouldnt be calling `onError` in case of successful response
```suggestion
```

Shiti commented on src/components/media/controls/videoControls.tsx:
```diff
@@ -0,0 +1,92 @@
+      props.element
+        .requestFullscreen()
+        .then(() => {
+          props.onError('')
```
We shouldnt be calling `onError` in case of successful response
```suggestion
```

Shiti commented on src/components/media/video/video.tsx:
```diff
@@ -0,0 +1,329 @@
+    return `rgba(${r}, ${g}, ${b}, ${opacity})`
+  }
+
+  const backgroundColor = adjustHexColorOpacity(
+    theme.palette.background.paper,
+    0.6
+  )
```
I think we should not set a opacity of 0.6 but leave it at 1. In case of videos, it can effect visibility since the video may be running and could have conflicting colors.

Shiti commented on src/components/media/controls/commonControls.tsx:
```diff
@@ -194,36 +160,45 @@ export function TranscriptToggle(props: TranscriptToggleProps) {
-      props.mediaElement.play()
-      setIsPlaying(true)
+      props.mediaElement.play().catch(() => {
+        props.onError('Failed to play the media. Please try again.')
```
You are catching the exception but not capturing any information regarding it. 
> Possible errors include:
> * [NotAllowedError](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play#notallowederror) [DOMException](https://developer.mozilla.org/en-US/docs/Web/API/DOMException)
Provided if the user agent (browser) or operating system doesn't allow playback of media in the current context or situation. The browser may require the user to explicitly start media playback by clicking a "play" button, for example because of a [Permissions Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Permissions_Policy).
>* [NotSupportedError](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play#notsupportederror) [DOMException](https://developer.mozilla.org/en-US/docs/Web/API/DOMException)
Provided if the media source (which may be specified as a [MediaStream](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream), [MediaSource](https://developer.mozilla.org/en-US/docs/Web/API/MediaSource), [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob), or [File](https://developer.mozilla.org/en-US/docs/Web/API/File), for example) doesn't represent a supported media format.

In these scenarios, trying again will be of no use. The appropriate message should be indicated clearly so user understands what the issue is and why the media cannot be played.

Shiti commented on src/components/media/video/video.css:
```diff
@@ -0,0 +1,108 @@
+
+.rustic-video-control-error-message {
+  position: absolute;
+  top: 0px;
+  left: 0px;
```
```suggestion
  top: 0;
  left: 0;
```

Shiti commented on src/components/media/video/video.tsx:
```diff
@@ -0,0 +1,336 @@
+/* eslint-disable no-magic-numbers */
```
```suggestion
```

Shiti commented on src/components/media/video/video.tsx:
```diff
@@ -0,0 +1,336 @@
+  const videoRef = useRef<HTMLVideoElement>(null)
+  const theme = useTheme()
+  const isMobile = useMediaQuery(theme.breakpoints.down('md'))
+
```
```suggestion
 const inActivityTimeout = 3000 // Hide controls after 3 seconds of inactivity
```

Shiti commented on src/components/media/video/video.tsx:
```diff
@@ -0,0 +1,336 @@
+
+    function startTimeout() {
+      clearTimeout(timeoutId)
+      timeoutId = setTimeout(() => setAreControlsVisible(false), 3000) // Hide controls after 3 seconds of inactivity
```
```suggestion
      timeoutId = setTimeout(() => setAreControlsVisible(false), inActivityTimeout) 
```
